### PR TITLE
feat: NOT_READY after provider shutdown

### PR DESCRIPTION
**Pull request by toddbaert**
Add requirement clarifying providers requiring initialization return to `NOT_READY` after shutdown, as agreed on [here](https://github.com/open-feature/spec/issues/213).

Fixes: https://github.com/open-feature/spec/issues/213

### CODE REVIEW

File: specification/sections/02-providers.md [hunk 1]
```diff
@@ -203,12 +203,15 @@ title: Provider State
...
-    NOT_READY --> READY
+    NOT_READY --> READY:initialize
     READY --> ERROR
     ERROR --> READY
     READY --> STALE
     STALE --> READY
     STALE --> ERROR
+    READY --> NOT_READY:shutdown
+    STALE --> NOT_READY:shutdown
+    ERROR --> NOT_READY:shutdown
```
Comment (toddbaert) [inline specification/sections/02-providers.md:214]: I added names for a few of the transitions because it made the diagram less confusing. Adding them everywhere seemed redundant though.

Comment (Kavindu-Dodan) [reply specification/sections/02-providers.md:214]: The preview looks good. IMO good improvement to explain _what_ cause the state change.

```mermaid
---
title: Provider State
---
stateDiagram-v2
    direction LR
    [*] --> NOT_READY
    NOT_READY --> READY:initialize
    READY --> ERROR
    ERROR --> READY
    READY --> STALE
    STALE --> READY
    STALE --> ERROR
    READY --> NOT_READY:shutdown
    STALE --> NOT_READY:shutdown
    ERROR --> NOT_READY:shutdown
```

File: specification/sections/02-providers.md [hunk 2]
```diff
@@ -250,6 +253,15 @@ class MyProvider implements Provider, AutoDisposable {
   }
 ```
 
+#### Requirement 2.5.2
+
+> After a provider's shutdown function has terminated successfully, the provider's state **MUST** revert to its uninitialized state.
+
+If a provider requires initialization, once it's shut down, it should transition to its initial `NOT_READY` state, so it can be reinitialized.
```
Comment (Kavindu-Dodan) [inline specification/sections/02-providers.md:260]: >  ... so it can be reinitialized

I think we should not enforce this behavior through spec 🤔

**Review by Kavindu-Dodan — APPROVED** [review]
Left [this](https://github.com/open-feature/spec/pull/216/files#r1394677422) comment, otherwise change looks fine.

File: specification/sections/02-providers.md (hunk 2 above)
Comment (toddbaert) [reply specification/sections/02-providers.md:260]: Do you mean we should just remove the "so it can be reinitialized"?

We could remove it, but I think that this is the logical implication of transitioning back to `NOT_READY`. On the other hand, I guess you could make the argument some providers could still be unable to be made `READY` again, and they could throw if they are reinitialized.

File: specification/sections/02-providers.md [hunk 3]
```diff
@@ -250,6 +253,15 @@ class MyProvider implements Provider, AutoDisposable {
   }
 ```
 
+#### Requirement 2.5.2
+
+> After a provider's shutdown function has terminated successfully, the provider's state **MUST** revert to its uninitialized state.
```
Comment (beeme1mr) [inline specification/sections/02-providers.md:258]: The non-normative section says "should," while the spec says" must".

File: specification/sections/02-providers.md (hunk 2 above)
Comment (toddbaert) [inline specification/sections/02-providers.md:260]:
```suggestion
If a provider requires initialization, once it's shut down, it must transition to its initial `NOT_READY` state, so it can be reinitialized.
```

File: specification/sections/02-providers.md (hunk 3 above)
Comment (toddbaert) [reply specification/sections/02-providers.md:258]: Good call, fixed.

File: specification/sections/02-providers.md (hunk 2 above)
Comment (Kavindu-Dodan) [reply specification/sections/02-providers.md]: Yes, originally thought we shouldn't define this through the spec. But with your explanation, I think we should reword it to have **MAY** - ex:-  ... so it **MAY** be reinitialized.

File: specification/sections/02-providers.md [hunk 4]
```diff
@@ -250,6 +253,15 @@ class MyProvider implements Provider, AutoDisposable {
...
+
+If a provider requires initialization, once it's shut down, it must transition to its initial `NOT_READY` state, so it can be reinitialized.
```
Comment (toddbaert) [inline specification/sections/02-providers.md:260]:
```suggestion
If a provider requires initialization, once it's shut down, it must transition to its initial `NOT_READY` state. Some providers may allow re-initialization from this state.
```

File: specification/sections/02-providers.md [hunk 5]
```diff
@@ -250,6 +253,15 @@ class MyProvider implements Provider, AutoDisposable {
   }
 ```
 
+#### Requirement 2.5.2
+
+> After a provider's shutdown function has terminated successfully, the provider's state **MUST** revert to its uninitialized state.
+
+If a provider requires initialization, once it's shut down, it must transition to its initial `NOT_READY` state. Some providers may allow re-initialization from this state.
```
Comment (toddbaert) [inline specification/sections/02-providers.md:260]:
```suggestion
If a provider requires initialization, once it's shut down, it must transition to its initial `NOT_READY` state. Some providers may allow reinitialization from this state.
```

**Review by lukas-reining — APPROVED** [review]
This looks good to me.

**Comment by toddbaert** [comment]
I'm going to merge this Friday EOD unless anyone has concerns.

**Review by weyert — APPROVED** [review]
